### PR TITLE
sqlparser: introduce custom formatting options

### PR DESCRIPTION
**Pull request by vmg**
## Description

@shlomi-noach has requested some more options to the way `sqlparser` serializes its output, so we can better match canonical MySQL output. This PR introduces flags to enable normalizing all SQL into uppercase, and to escape all identifiers in a query.

cc @systay @GuptaManan100 for review

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

### CODE REVIEW

**Review by shlomi-noach — COMMENTED** [review]
This looks amazing! I will add a few more test statements/scenarios and see if there's anything else to flesh.

cc @iheanyi

**Comment by vmg** [comment]
@shlomi-noach: the parsing test are failing on definers, as there is double escaping now. What do you think is the right way here? Is the original query with `'create'` valid, or should it have been using backticks to begin with?

```
        Expected/Got:
        create algorithm = temptable definer = 'create'@b view a(b, c, d) as select * from e with local check option
        create algorithm = temptable definer = `'create'`@b view a(b, c, d) as select * from e with local check option
```

**Comment by vmg** [comment]
The MySQL docs say this:

> If the DEFINER clause is present, the user value should be a MySQL account specified as 'user_name'@'host_name', [CURRENT_USER](https://dev.mysql.com/doc/refman/8.0/en/information-functions.html#function_current-user), or [CURRENT_USER()](https://dev.mysql.com/doc/refman/8.0/en/information-functions.html#function_current-user). The permitted user values depend on the privileges you hold, as discussed in [Section 25.6, “Stored Object Access Control”](https://dev.mysql.com/doc/refman/8.0/en/stored-objects-security.html). Also see that section for additional information about stored routine security.

But the example you showed in Slack was using backticks for the username. How did you generate that example?

**Comment by vmg** [comment]
OK I've fixed the `definer` behavior to do what we were previously doing, but making the identifier case-sensitive so it doesn't get upcased. This passes all tests and seems to match MySQL's behavior much closer (modulo the example you pasted in Slack, which still appears weird to me -- all the other examples I've seen online expect single ticks).

**Comment by shlomi-noach** [comment]
> Is the original query with 'create' valid, or should it have been using backticks to begin with?

Yeah - completely valid, and even the documented way of doing so! See https://dev.mysql.com/doc/refman/8.0/en/create-user.html:

```
CREATE USER 'jeffrey'@'localhost'
  IDENTIFIED BY 'new_password' PASSWORD EXPIRE;
```

**Comment by shlomi-noach** [comment]
I've modified the test a bit in [b74872d89a](https://github.com/vitessio/vitess/pull/10068/commits/b74872d89a1ccc0c0898e815f8e9b95d1e8ba326) as follows:

- after we've evaluated the `out` -- the serialized formatted query,
- we re-parse it to validate there's no parsing errors (so we know we generated a valid statement)
- and re-serialize the re-parsed statement to validate we still produce the exact same query

**Review by GuptaManan100 — APPROVED** [review]
All the other changes look great to me!

**Review by shlomi-noach — COMMENTED** [review]
Should we introduce a few function, say `sqlparser.CanonicalString(stmt)`, similar to `ssqlparser.String(stmt)`, that wraps up the new behavior like so:

```go
func CanonicalString(stmt Statement) string {
	buf := NewTrackedBuffer(nil)
	if flags.escape {
		buf.SetEscapeAllIdentifiers(true)
	}
	if flags.upcase {
		buf.SetUpperCase(true)
	}
	buf.Myprintf("%v", stmt)

	out := buf.String()
	return out
}
```

?

File: go/vt/sqlparser/tracked_buffer.go [hunk 1]
```diff
@@ -98,14 +138,20 @@ func (buf *TrackedBuffer) astPrintf(currentNode SQLNode, format string, values .
...
 		}
 		i++ // '%'
-		token := format[i]
-		switch token {
+
+		alternate := false
+		if format[i] == '#' {
+			alternate = true
+			i++
+		}
```
[the remark this answers is not in the document]
Comment (vmg) [reply go/vt/sqlparser/tracked_buffer.go:152]: That's right, I'll give this a more reasonable name.

**Review by shlomi-noach — APPROVED** [review]
Please make sure to Squash and merge, in case we want to backport this PR

This is so good, thank you very much! :heart:

**Comment by vmg** [comment]
Yikes, forgot to squash merge. Sure wish it could be made the default. :sleepy:

**Pull request merged**